lang/cs/Org.Apache.REEF.Network/Group/Driver/Impl/ContextManager.cs

```diff
@@ -0,0 +1,148 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Org.Apache.REEF.Driver.Context;
+using Org.Apache.REEF.Utilities.Diagnostics;
+using Org.Apache.REEF.Utilities.Logging;
+
+namespace Org.Apache.REEF.Network.Group.Driver.Impl
+{
+    /// <summary>
+    /// This class is to manage the active contexts in group communications
+    /// </summary>
+    public class ContextManager
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(ContextManager));
+        private readonly ReaderWriterLockSlim _contextLock = new ReaderWriterLockSlim();
+        private readonly int _totalNumberOfContexts;
+        private readonly IDictionary<string, IActiveContext> _activeContexts;
+
+        public ContextManager(int numberOfContexts)
+        {
+            _totalNumberOfContexts = numberOfContexts;
+            _activeContexts = new Dictionary<string, IActiveContext>();
+        }
+
+        /// <summary>
+        /// Check if the number of active context has reached to the total number
+        /// </summary>
+        /// <returns></returns>
+        public bool AllContextReceived()
+        {
+            lock (_contextLock)
```

Comment:
I do not understand the use here....you are using both lock and _contextLock simultaneously?? Why?